theano/gpuarray/opt.py

```diff
@@ -210,6 +210,52 @@ def apply(self, fgraph):
                     0, 'fast_run', 'fast_compile', 'merge')
 
 
+class GraphToGPU(Optimizer):
+    """
+    Transfer the graph as a whole to GPU instead of replacing nodes
+    """
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        move_to_GPU = True
+
+        # Building a new graph
+        for i in fgraph.inputs:
+            mapping[i] = GpuFromHost(None)(i)
```

Comment:
That was the confirmation that I wanted. If that is the case, then in the dict `mapping`, we're creating the dictionary with inputs to the graph, and while  we're cloning the node, we're using the inputs to an apply node as the key. I don't think that would be the same ?
Also, as a workaround, can we do something like this ?
 
`
for n in fgraph.toposort():     
    mapping[n.inputs] = GpuFromHost(None)(n.inputs)`